### ISSUE TITLE
Extremely leaky Perl 5.8.0

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#5949]
Migrated from [rt.perl.org#17460](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=17460) (status was 'resolved')

Searchable as RT17460$

**Comment by p5pRT** [comment]
### From efinley@smfbsd.etv.net

#### Created by efinleywork@efinley.com



This program doesn't finish because it runs out of memory&#8203;:

\#\!/usr/bin/perl \-w

use strict;
use Config;
use threads;
use threads&#8203;::shared;
use Thread&#8203;::Queue;

$Config\{usethreads\} or die "Recompile Perl with threads to run this program\.";

main\(\);
exit;

sub main \{
&nbsp;    my $q = new Thread&#8203;::Queue;
&nbsp;    my $t = threads\->create\(\\&test\, $q\);

&nbsp;    my $line;
&nbsp;    foreach my $i \(1\.\.1000000\) \{
&nbsp;        $q\->enqueue\($i\);
&nbsp;    \}

&nbsp;    $q\->enqueue\(undef\);

&nbsp;    $t\->join;
\}

sub test \{
&nbsp;    my \($in\) = @&#8203;\_;

&nbsp;    my $num;
&nbsp;    while\(defined\($num = $in\->dequeue\(\)\)\) \{
&nbsp;        $num = undef;
&nbsp;    \}
\}

<details><summary>Perl Info</summary>

```


This perlbug was built using Perl 5.00503 - Sun Mar  5 13:39:27 SAST 2000
It is being executed now by  Perl 5.008 - Wed Sep 11 20:29:31 MDT 2002.

Site configuration information for perl 5.008:

Configured by efinley at Wed Sep 11 20:29:31 MDT 2002.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=freebsd, osvers=4.6-stable, archname=i386-freebsd-thread-multi
    uname='freebsd smfbsd.etv.net 4.6-stable freebsd 4.6-stable #0: wed aug 28 14:08:40 mdt 2002 root@smfbsd.etv.net:usrobjusrsrcsyssmfbsd i386 '
    config_args='-sde -Dprefix=/usr/local -Darchlib=/usr/local/lib/perl5/5.8.0/mach -Dprivlib=/usr/local/lib/perl5/5.8.0 -Dman3dir=/usr/local/lib/perl5/5.8.0/man/man3 -Dsitearch=/usr/local/lib/perl5/site_perl/5.8.0/mach -Dsitelib=/usr/local/lib/perl5/site_perl/5.8.0 -Ui_malloc -Ui_iconv -Uinstallusrbinperl -Dccflags=-DAPPLLIB_EXP="/usr/local/lib/perl5/5.8.0/BSDPAN" -Dusethreads -Ui_gdbm -Dusemymalloc=y'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-DAPPLLIB_EXP="/usr/local/lib/perl5/5.8.0/BSDPAN" -DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -fno-strict-aliasing -I/usr/local/include',
    optimize='-O -pipe ',
    cppflags='-DAPPLLIB_EXP="/usr/local/lib/perl5/5.8.0/BSDPAN" -DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='2.95.4 20020320 [FreeBSD]', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags ='-pthread -Wl,-E  -L/usr/local/lib'
    libpth=/usr/lib /usr/local/lib
    libs=-lgdbm -lm -lc_r -lcrypt -lutil
    perllibs=-lm -lc_r -lcrypt -lutil
    libc=, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-DPIC -fPIC', lddlflags='-shared  -L/usr/local/lib'

Locally applied patches:
    


@INC for perl 5.008:
    /usr/local/lib/perl5/site_perl/5.8.0/mach
    /usr/local/lib/perl5/site_perl/5.8.0
    /usr/local/lib/perl5/site_perl
    /usr/local/lib/perl5/5.8.0/BSDPAN
    /usr/local/lib/perl5/5.8.0/mach
    /usr/local/lib/perl5/5.8.0
    .


Environment for perl 5.008:
    HOME=/home/efinley
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/sbin:/bin:/usr/sbin:/usr/bin:/usr/games:/usr/local/sbin:/usr/local/bin:/usr/X11R6/bin:/home/efinley/bin
    PERL_BADLANG (unset)
    SHELL=/bin/csh

```
</details>

**Comment by p5pRT** [comment]
### From @lizmat

At 10&#8203;:27 PM 9/20/02 \+0000\, efinley@&#8203;smfbsd\.etv\.net \(via RT\) wrote&#8203;:
>This program doesn't finish because it runs out of memory&#8203;:
>use threads&#8203;::shared;
>use Thread&#8203;::Queue;

This is a known problem with shared arrays \(and possibly shared 
hashes\)\.  Arthur Bergman is working on this and expects to release a new 
version of threads\.pm and/or threads&#8203;::shared\.pm on CPAN Real Soon Now \(once 
he is recovered from the YAPC&#8203;::Europe\)\.

\<plug author="me">
In the meantime\, you could try Thread&#8203;::Conveyor from CPAN which doesn't 
leak \(because it doesn't use shared arrays\) but may be slower because it's 
all done in Perl \(still\)\.  On the other hand\, it has some features 
Thread&#8203;::Queue doesn't have\, such as throttling \(blocking the enqueueing 
thread if there are too many entries in the queue\) and monitoring\.
\</plug>


Liz

**Comment by p5pRT** [comment]
### From jim.cistaro@av.com

I replied to \#15893 but did not get any response from anyone so I will
try this one\.

from message msg86928
 
I am a relative novice as XS so please pardon my ignorance\.  I
experience the same memory leak issues stated in bug 15893 while trying
to use Threads&#8203;::Queue \(hence threads&#8203;::shared\)\.  After looking at the
code generated in shared\.c and reading up on the documentation\, I had a
few questions\.

Is there a reason that POP and unshift use 
ST\(0\) = Nullsv;
instead of
ST\(0\) = sv\_newmortal\(\);

Also should they do a 
SvREFCNT\_dec\(sv\);
This would appear to reduce the refcount to 1 if it is a normal scalar\. 
I am guessing that 1 is associated with the mortal ST\(0\)\.  Then freeing
occurs when ST\(0\) dies\.

I have only done minimal testing of a simple queue\, but this seems to
stop the leak\.  I believe the only growth I see is due to the queue
backing up\.

As I said\, I am somewhat of a novice at XS and just want to be able to
use threading with queues\.  Hope this helps\.  If not\, sorry for taking
up anyone's time\.

Jim


On Thu\, 1999\-12\-30 at 19&#8203;:00\, efinley@&#8203;smfbsd\.etv\.net \(via RT\) wrote&#8203;:
> \# New Ticket Created by  efinley@&#8203;smfbsd\.etv\.net 
> \# Please include the string&#8203;:  \[perl \#17460\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt2/Ticket/Display\.html?id=17460 >
> 
> 
> 
> This is a bug report for perl from efinleywork@&#8203;efinley\.com\,
> generated with the help of perlbug 1\.26 running under perl 5\.008\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> This program doesn't finish because it runs out of memory&#8203;:
> 
> \#\!/usr/bin/perl \-w
> 
> use strict;
> use Config;
> use threads;
> use threads&#8203;::shared;
> use Thread&#8203;::Queue;
> 
> $Config\{usethreads\} or die "Recompile Perl with threads to run this program\.";
> 
> main\(\);
> exit;
> 
> sub main \{
>     my $q = new Thread&#8203;::Queue;
>     my $t = threads\->create\(\\&test\, $q\);
> 
>     my $line;
>     foreach my $i \(1\.\.1000000\) \{
>         $q\->enqueue\($i\);
>     \}
> 
>     $q\->enqueue\(undef\);
> 
>     $t\->join;
> \}
> 
> sub test \{
>     my \($in\) = @&#8203;\_;
> 
>     my $num;
>     while\(defined\($num = $in\->dequeue\(\)\)\) \{
>         $num = undef;
>     \}
> \}
> 
> 
> 
> \[Please do not change anything below this line\]
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> 
> \-\-\-
> This perlbug was built using Perl 5\.00503 \- Sun Mar  5 13&#8203;:39&#8203;:27 SAST 2000
> It is being executed now by  Perl 5\.008 \- Wed Sep 11 20&#8203;:29&#8203;:31 MDT 2002\.
> 
> Site configuration information for perl 5\.008&#8203;:
> 
> Configured by efinley at Wed Sep 11 20&#8203;:29&#8203;:31 MDT 2002\.
> 
> Summary of my perl5 \(revision 5\.0 version 8 subversion 0\) configuration&#8203;:
>   Platform&#8203;:
>     osname=freebsd\, osvers=4\.6\-stable\, archname=i386\-freebsd\-thread\-multi
>     uname='freebsd smfbsd\.etv\.net 4\.6\-stable freebsd 4\.6\-stable \#0&#8203;: wed aug 28 14&#8203;:08&#8203;:40 mdt 2002 root@&#8203;smfbsd\.etv\.net&#8203;:usrobjusrsrcsyssmfbsd i386 '
>     config\_args='\-sde \-Dprefix=/usr/local \-Darchlib=/usr/local/lib/perl5/5\.8\.0/mach \-Dprivlib=/usr/local/lib/perl5/5\.8\.0 \-Dman3dir=/usr/local/lib/perl5/5\.8\.0/man/man3 \-Dsitearch=/usr/local/lib/perl5/site\_perl/5\.8\.0/mach \-Dsitelib=/usr/local/lib/perl5/site\_perl/5\.8\.0 \-Ui\_malloc \-Ui\_iconv \-Uinstallusrbinperl \-Dccflags=\-DAPPLLIB\_EXP="/usr/local/lib/perl5/5\.8\.0/BSDPAN" \-Dusethreads \-Ui\_gdbm \-Dusemymalloc=y'
>     hint=recommended\, useposix=true\, d\_sigaction=define
>     usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
>     useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
>     use64bitint=undef use64bitall=undef uselongdouble=undef
>     usemymalloc=n\, bincompat5005=undef
>   Compiler&#8203;:
>     cc='cc'\, ccflags ='\-DAPPLLIB\_EXP="/usr/local/lib/perl5/5\.8\.0/BSDPAN" \-DHAS\_FPSETMASK \-DHAS\_FLOATINGPOINT\_H \-fno\-strict\-aliasing \-I/usr/local/include'\,
>     optimize='\-O \-pipe '\,
>     cppflags='\-DAPPLLIB\_EXP="/usr/local/lib/perl5/5\.8\.0/BSDPAN" \-DHAS\_FPSETMASK \-DHAS\_FLOATINGPOINT\_H \-fno\-strict\-aliasing \-I/usr/local/include'
>     ccversion=''\, gccversion='2\.95\.4 20020320 \[FreeBSD\]'\, gccosandvers=''
>     intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
>     d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
>     ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
>     alignbytes=4\, prototype=define
>   Linker and Libraries&#8203;:
>     ld='cc'\, ldflags ='\-pthread \-Wl\,\-E  \-L/usr/local/lib'
>     libpth=/usr/lib /usr/local/lib
>     libs=\-lgdbm \-lm \-lc\_r \-lcrypt \-lutil
>     perllibs=\-lm \-lc\_r \-lcrypt \-lutil
>     libc=\, so=so\, useshrplib=false\, libperl=libperl\.a
>     gnulibc\_version=''
>   Dynamic Linking&#8203;:
>     dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags=' '
>     cccdlflags='\-DPIC \-fPIC'\, lddlflags='\-shared  \-L/usr/local/lib'
> 
> Locally applied patches&#8203;:
>     
> 
> \-\-\-
> @&#8203;INC for perl 5\.008&#8203;:
>     /usr/local/lib/perl5/site\_perl/5\.8\.0/mach
>     /usr/local/lib/perl5/site\_perl/5\.8\.0
>     /usr/local/lib/perl5/site\_perl
>     /usr/local/lib/perl5/5\.8\.0/BSDPAN
>     /usr/local/lib/perl5/5\.8\.0/mach
>     /usr/local/lib/perl5/5\.8\.0
>     \.
> 
> \-\-\-
> Environment for perl 5\.008&#8203;:
>     HOME=/home/efinley
>     LANG \(unset\)
>     LANGUAGE \(unset\)
>     LD\_LIBRARY\_PATH \(unset\)
>     LOGDIR \(unset\)
>     PATH=/sbin&#8203;:/bin&#8203;:/usr/sbin&#8203;:/usr/bin&#8203;:/usr/games&#8203;:/usr/local/sbin&#8203;:/usr/local/bin&#8203;:/usr/X11R6/bin&#8203;:/home/efinley/bin
>     PERL\_BADLANG \(unset\)
>     SHELL=/bin/csh
> 
>

**Comment by p5pRT** [comment]
arthur@contiller.se - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From arthur@contiller.se

Fixed by change \#19200\, same as bug 15983

**Comment by p5pRT** [comment]
arthur@contiller.se - Status changed from 'open' to 'resolved'
